Sources/Core/Tracker/Tracker.swift

```diff
@@ -385,7 +385,81 @@ class Tracker: NSObject {
         emitter.resumeTimer()
         session?.startChecker()
     }
-
+    
+    /// - Returns: The associated value of the extended parameter if enabled and set in the tracker, an empty string if the parameter is not enabled, or nil if parameter is enabled without an aviliable value
+    func tryExtendedParameter(_ extendedParameterSet: Bool, _ value: String?) -> String? {
+        switch (extendedParameterSet, value) {
+        case (true, .some(let val)):
+            return val
+        case (true, .none):
+            return nil
+        case (false, _):
+            return ""
+        }
+    }
+    
+    func decorateLinkErrorTemplate(_ extendedParameterName: String) -> String {
+        "Cannot decorate link: \(extendedParameterName) has been requested in CrossDeviceParameterConfiguration, but it is not set."
+    }
+    
+    @objc func decorateLink(_ url: URL, extendedParameters: CrossDeviceParameterConfiguration? = nil) -> URL? {
+        var userId: String
+        switch self.session {
+        case .none:
+            logError(message: "\(url) could not be decorated as session.userId is nil")
+            return nil
+        case .some(let s):
+            userId = s.userId
+        }
+        
+        let extendedParameters = extendedParameters ?? CrossDeviceParameterConfiguration()
+        
+        guard let sessionId = tryExtendedParameter(extendedParameters.sessionId, self.session?.state?.sessionId) else {
+            logError(message: "\(decorateLinkErrorTemplate("sessionId")) Ensure an event has been tracked to generate a session before calling this method.")
```

Comment:
I don't think this matches the behaviour in the Android tracker – if some requested properties in there are not available, they are set to an empty string. I think I prefer the Android tracker implementation, this seems a bit too strict to return nothing in case a configured parameter is not available (except when dealing with the required session.userId). It puts a larger burden on the user to have to deal with all the possible missing properties and handle them (if at all possible because we don't return info about which parameter is missing from this function).

Reply:
Updated to set to empty string and log debug messages if missing.